app/controllers/application_controller.rb

```diff
@@ -187,8 +188,11 @@ def expanded_embed(url, caption=nil)
       end
     end
 
-    render_to_string partial: "/articles/embeds/#{slug}.html.erb", locals: { embed_id: embed_id || url, caption: caption }
+    render_to_string partial: "/articles/embeds/#{slug}.html.erb", locals: { embed_id: embed_id || url, caption: caption, link: link }
   end
   helper_method :expanded_embed
 
+  def url_or_path?(string)
+    string.match?(/^(http|\/)\S+/)
```

Comment:
It covers the first 3 cases.  I can modify it to cover the final case, but that seems like it would have the most false positives.  I can see wanting to have a caption with a / in it quite often.